collector/logs/sources/tail/tailer.go

```diff
@@ -0,0 +1,201 @@
+package tail
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"time"
+
+	"github.com/Azure/adx-mon/collector/logs/engine"
+	"github.com/Azure/adx-mon/collector/logs/sources/tail/sourceparse"
+	"github.com/Azure/adx-mon/collector/logs/transforms/parser"
+	"github.com/Azure/adx-mon/collector/logs/types"
+	"github.com/Azure/adx-mon/pkg/logger"
+	"github.com/tenebris-tech/tail"
+	"golang.org/x/sync/errgroup"
+)
+
+type TailerConfig struct {
+	Target          FileTailTarget
+	UpdateChan      <-chan FileTailTarget
+	AckGenerator    func(*types.Log) func()
+	WorkerCreator   engine.WorkerCreatorFunc
+	CursorDirectory string
+	WorkerName      string
+}
+
+// Tailer is a specific instance of a file being tailed.
+type Tailer struct {
+	tail           *tail.Tail
+	shutdown       context.CancelFunc
+	errgroup       *errgroup.Group
+	database       string
+	table          string
+	logTypeParser  sourceparse.LogTypeParser
+	logLineParsers []parser.Parser
+	resources      map[string]interface{}
+}
+
+func StartTailing(config TailerConfig) (*Tailer, error) {
+	group, ctx := errgroup.WithContext(context.Background())
+	ctx, shutdown := context.WithCancel(ctx)
+
+	batchQueue := make(chan *types.Log, 512)
+	outputQueue := make(chan *types.LogBatch, 1)
+	tailConfig := tail.Config{Follow: true, ReOpen: true, Poll: true}
+	existingCursorPath := cursorPath(config.CursorDirectory, config.Target.FilePath)
+	fileId, position, err := readCursor(existingCursorPath)
+	if err == nil {
+		if logger.IsDebug() {
+			logger.Debugf("TailSource: found existing cursor for file %q: %s %d", config.Target.FilePath, fileId, position)
+		}
+		tailConfig.Location = &tail.SeekInfo{
+			Offset:         position,
+			Whence:         io.SeekStart,
+			FileIdentifier: fileId,
+		}
+	}
+
+	tailFile, err := tail.TailFile(config.Target.FilePath, tailConfig)
+	if err != nil {
+		shutdown()
+		return nil, fmt.Errorf("addTarget create tailfile: %w", err)
+	}
+
+	parsers, err := parser.NewParsers(config.Target.Parsers)
+	if err != nil {
+		shutdown()
+		return nil, fmt.Errorf("addTarget create parsers: %w", err)
+	}
+
+	attributes := make(map[string]interface{})
+	for k, v := range config.Target.Resources {
+		attributes[k] = v
+	}
+
+	tailer := &Tailer{
+		tail:           tailFile,
+		shutdown:       shutdown,
+		errgroup:       group,
+		database:       config.Target.Database,
+		table:          config.Target.Table,
+		logTypeParser:  sourceparse.GetLogTypeParser(config.Target.LogType),
+		logLineParsers: parsers,
+		resources:      attributes,
+	}
+	group.Go(func() error {
+		return readLines(ctx, tailer, config.UpdateChan, batchQueue)
```

Comment:
Along the same thought process as the previous comment, it seems that returning an error into the error group due to what could easily be a transitive failure, causing us to abandon tailing this file, is something we're going to want to take a closer look at in the future. I think we should error on the side of robustness / permissive error handling when tailing these files such that we integrate a reasonable retry strategy into any error that is transitive in nature; or maybe even treat all errors that are transitive except for a select few that we know are permanent, like `os.ErrNotExist`.